### PR TITLE
doc: Add a spec for Application State

### PR DESCRIPTION
**Pull request by DHowett**
This draft pull request ([Review link](https://github.com/microsoft/terminal/blob/dev/duhowett/spec/appstate/doc/specs/%23xxxx%20-%20Application%20State%20(TSM).md)) introduces a specification for "cross-session" app state that isn't stored in the user's settings.json.

## PR Checklist
* [x] Specs #8324
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed
* [x] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [x] Schema updated.
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

[Review link](https://github.com/microsoft/terminal/blob/dev/duhowett/spec/appstate/doc/specs/%238324%20-%20Application%20State%20(TSM).md)

### CODE REVIEW

File: doc/specs/#xxxx - Application State (TSM).md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+### Security
+
+It is not expected that this proposal will impact security, as it uses our _existing_ JSON parser in a new place.
+
+### Reliability
+
+[comment]: # Will the proposed change improve reliability? If not, why make the change?
+
+The comment in this section heavily suggests that we should only make changes that improve reliability, not ones that
+maintain it.
```
Comment (zadjii-msft) [inline doc/specs/#xxxx - Application State (TSM).md:76]: \<aside> Yea, this "Capabilities" section I've always hated. I think if you're outside MSFT, then you don't have any idea what to put here, and "capabilities" is a _terrible_ heading for it

File: doc/specs/#xxxx - Application State (TSM).md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+place.
+
+Of note: there is no explicit `Save` or `Commit` mechanism. Changes to the application state are committed durably a
+short duration after they're made.
+
+## UI/UX Design
+
+This has no direct impact on the UI/UX; however, we may want to add a button to general settings page titled "reset all
+dialogs" or "don't not ask me again about those things that, some time ago, I told you to not ask me about".
+
```
Comment (zadjii-msft) [inline doc/specs/#xxxx - Application State (TSM).md:60]: I'd maybe add a note that we're not intending for the user to hand-edit `state.json`, so we're not as concerned about how the contents of that file are serialized. This is unlike `settings.json`, which has to have a semblance of user friendliness.

**Review by zadjii-msft — APPROVED** [review]
Yea this all seems reasonable to me.

File: doc/specs/#xxxx - Application State (TSM).md (hunk 1 above)
Comment (DHowett) [reply doc/specs/#xxxx - Application State (TSM).md:76]: _agree_

File: doc/specs/#xxxx - Application State (TSM).md (hunk 2 above)
Comment (DHowett) [reply doc/specs/#xxxx - Application State (TSM).md:60]: Good call. Thanks!

**Review by carlos-zamora — APPROVED** [review]
More curious about the applications than the overall design, itself. But it's fair to say that that's out of scope so meh.

**Comment by DHowett** [comment]
@carlos-zamora clarify? A whole section was dedicated to _why_ (the applications). Any questions I can answer?

**Comment by carlos-zamora** [comment]
> @carlos-zamora clarify? A whole section was dedicated to _why_ (the applications). Any questions I can answer?

I don't mean it as a bad thing haha. I mean more that I'm curious what the design for "don't ask me again" or "keep the profile deleted" would look like. But I'm guessing the specific design for those settings is a bit out of scope here (if they even need a design, frankly).

Like, I'm curious if we should move `disabledProfileSources` over or how that would interact with the plan for "keep the dynamic profile deleted"? But that's probably out of scope.

**Comment by DHowett** [comment]
So, `disabledSources` is twofold:

1. It means “do not run the expensive profile search code” (like the VS Generator!)
2. When extensions exist, they could be suppressed.

Disabling the source is a bigger hammer than deleting one profile generated by the source.

**Comment by DHowett** [comment]
Reasonable question though!

File: doc/specs/#8324 - Application State (TSM).md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+> 📝 The sole motivation for exposing this from Microsoft.Terminal.Settings is to concentrate JSON de-/serialization in one
+place.
+
+Of note: there is no explicit `Save` or `Commit` mechanism. Changes to the application state are committed durably a
+short duration after they're made.
+
+## UI/UX Design
+
+This has no direct impact on the UI/UX; however, we may want to add a button to general settings page titled "reset all
```
Comment (miniksa) [inline doc/specs/#8324 - Application State (TSM).md:58]: I think this is pretty important and we should ensure there's an easy reset.

**Review by miniksa — APPROVED** [review]
Looks fine to me.

**Pull request merged**